### PR TITLE
add deprecation warnings for old backends

### PR DESCRIPTION
**Pull request by michaelosthege**
Some old backends are cluttering the codebase. We don't know of anybody still using them, so this PR adds deprecation warnings.

I'll add a line to the release notes just in case anybody ignores warnings..

### CODE REVIEW

File: RELEASE-NOTES.md [hunk 1]
```diff
@@ -19,6 +19,7 @@
 - Remove `sample_ppc` and `sample_ppc_w` that were deprecated in 3.6.
 - Tuning results no longer leak into sequentially sampled `Metropolis` chains (see #3733 and #3796).
 - Deprecated `sd` in version 3.7 has been replaced by `sigma` now raises `DepreciationWarning` on using `sd` in continuous, mixed and timeseries distributions. (see #3837 and #3688).
```
Comment (ColCarroll) [inline RELEASE-NOTES.md]:
```suggestion
- Deprecated `sd` in version 3.7 has been replaced by `sigma` now raises `DeprecationWarning` on using `sd` in continuous, mixed and timeseries distributions. (see #3837 and #3688).
```

**Review by ColCarroll — APPROVED** [review]
I am, in general, super in favor of this!

My only reservation is that `save_trace`/`load_trace` is currently the only non-pickle way of reloading a trace to, for example, do ppc sampling. I'm sort of ok going back to suggesting pickle unless someone has a reasonable security reason not to.

**Comment by michaelosthege** [comment]
> I am, in general, super in favor of this!
> My only reservation is that save_trace/load_trace is currently the only non-pickle way of reloading a trace to, for example, do ppc sampling. I'm sort of ok going back to suggesting pickle unless someone has a reasonable security reason not to.

`sample_posterior_predictive` works with `InferenceData` since a few weeks ago.

**Comment by ColCarroll** [comment]
:heart_eyes_cat: deprecate and merge away, then!

**Comment by rpgoldman** [comment]
If we are going to deprecate save and load trace, shouldn't we go the further step of making `sample()` return an `InferenceData` object?

It seems like we shouldn't leave things in a state where we deprecate saving and loading `MultiTrace` objects, but still return them.

It also seems like we should revisit @aseyboldt 's idea of making PyMC3 variable creation support a `dims` argument (and possibly store `coords` on the `Model` object).

**Comment by michaelosthege** [comment]
I'm all in for returning `InferenceData`. I still find the `xarray` indexing super confusing, but _this is the way_.

I'll do a PR tomorrow to add `sample(..., return_inferencedata=False)`, with a `FutureWarning`, so we can change it to `sample(..., return_inferencedata=True)` in `v4.0.0`.
Then we'll see how much breaks (half the test suite I suppose 😅).

**Pull request merged**